### ISSUE TITLE
Error serializing libff::alt_bn128_Fr during DKG

### ISSUE THREAD

**Issue by gaoweihe** [skalenetwork/libBLS#174]
I am trying to distribute secret shares in the context of DKG. 
My plan is to serialize the `libff::alt_bn128_Fr` objects that `DKGBLSWrapper::createDKGSecretShares` produces, and distribute them to other participants through TCP network.

However, I got linking errors with the overloaded `operator<<` of `libff::alt_bn128_Fr`. 

The code segment relevant to this issue is

```
DKGBLSWrapper dkg_obj(t, n);

// serialize public shares
std::shared_ptr< std::vector< libff::alt_bn128_G2 > > public_shares = dkg_obj.createDKGPublicShares();
std::ostringstream oss_a;
oss_a << public_shares->at(0); // no error reported

// serialize private shares
std::shared_ptr< std::vector< libff::alt_bn128_Fr > > private_shares = dkg_obj.createDKGSecretShares();
std::ostringstream oss_b;
oss_b << private_shares->at(0); // FIXME: undefined reference to overloaded operator
```

The error info is 
```
    in function `std::ostream& libff::operator<< <4l>(std::ostream&, libff::bigint<4l> const&)': 
    /libff/algebra/fields/bigint.tcc:192: undefined reference to `operator<<(std::ostream&, __mpz_struct const*)'
```

My linking configs of cmake is 
```
target_link_libraries(project_name
    libsqlite3.so
    libmysqlcppconn.so
    libsodium.a
    libbls.a
    libff.a
    libgmp.so
    libgcov.so
    libssl.so
    libcrypto.a
    libz.so
    libmuduo_net.a
    libmuduo_base.a
    libpthread.so
    libdl.so
)
```

and cmake reported no errors during generation. 

Could you help me with this problem? 
In other words, what is the correct way to serialize and distribute public and private shares during DKG? Thanks.

┆Issue is synchronized with this [Jira Task](https://skalelabs.atlassian.net/browse/SKALE-5021)

**Comment by gaoweihe** [comment]
@olehnikolaiev

**Comment by olehnikolaiev** [comment]
hello @LuminadiaQ ! have you tried linking `gmp` library and others as static libraries not dynamic?

**Comment by gaoweihe** [comment]
Hi @olehnikolaiev . Thanks for your quick reply. 

Yes. I have tried that, but unfortunately it does not work. 

Now my linking configs is (`pthread` and `dl` are kept dynamic as otherwise other errors will emerge) 

```
target_link_libraries(tomchain-clt
    libsqlite3.a
    libmysqlcppconn.a
    libsodium.a
    libbls.a
    libff.a
    libgmp.a
    libgcov.a
    libssl.a
    libcrypto.a
    libz.a
    libmuduo_net.a
    libmuduo_base.a
    libpthread.so
    libdl.so
)
```

But I get the same error info 

```
in function `std::ostream& libff::operator<< <4l>(std::ostream&, libff::bigint<4l> const&)':
/libff/algebra/fields/bigint.tcc:192: undefined reference to `operator<<(std::ostream&, __mpz_struct const*)'
```

Is there a workaround for serialization? Thanks.

**Comment by gaoweihe** [comment]
Problem solved. Turns out I forgot to add the C++ library `gmpxx` together with `gmp`, which `ff` depends on. 

Thanks for your suggestions anyway, @olehnikolaiev . Your opinion has helped a lot.
